[DO NOT MERGE] Implement automatic merging of pull requests opened by JuliaRegistrator

DilumAluthge opened the pull request:
Fixes #3673 

## Summary

This is an implementation of automatic merging of pull requests made to a Julia package registry. It is suitable for use with the General registry as well as any other registry.

## Description

If a pull request is opened by the @JuliaRegistrator user, and if the pull request meets all of the automatic merging guidelines in #3673, then the Travis job will automatically merge the pull request.

If a pull request is not opened by the @JuliaRegistrator user, then it will not be automatically merged.

In order to enable this functionality, the following three steps need to be taken by someone with administrator permissions on the General registry:
- [ ] 1. Add a Travis encrypted environment variable with name `GITHUB_AUTOMERGE_TOKEN` that contains a GitHub token for a GitHub account that has push access to the General registry.
- [ ] 2. Add a Travis cron job for the General registry. The branch should be `master`. The interval should be `Daily`. The "options" should be `Always run`. **This last part is very important - the Travis cron job MUST be set to `Always run`. The merging will only happen on a Travis cron build or a Travis API build. Merging does not happen on regular Travis branch builds on the `master` branch. Therefore, if the Travis cron job is not set to `Always run`, then you will never automerge any pull requests!**
- [ ] 3. Merge this pull request into `master`.

I have set up a demo registry (https://github.com/bcbi/General). Please make pull requests to the demo registry to try this functionality out! Let me know if you want me to add your GitHub username to the list of usernames for which auto-merging is authorized.

---

## Note

It is important to note that if your pull request does not meet the guidelines for automatic merging, this does not mean that your pull request will never be merged. It just means that your pull request will require manual review by a human.

> These guidelines are intended not as requirements for packages but as very conservative guidelines, which, if your new package or new version of a package meets them, it may be automatically merged.

fredrikekre commented on .travis.yml:
```diff
@@ -9,7 +9,7 @@ julia:
 #   allow_failures:
 #     - julia: nightly
-script: julia .test/runtests.jl
+script: julia .test/runtests.jl && julia --project=.test/AutoMerge -e 'import Pkg; Pkg.instantiate()' && julia --project=.test/AutoMerge -e 'import Pkg; Pkg.test()' && julia --project=.test/AutoMerge -e 'using AutoMerge; AutoMerge.travis(verbose=true, approve_new_packages=true, merge_new_packages=true, approve_new_versions=true, merge_new_versions=true)'
```
How about pushing in `after_success`?

DilumAluthge replied:
Done!

fredrikekre commented on .travis.yml:
```diff
@@ -3,13 +3,14 @@ language: julia
-  # - 1.1
+  - "1.0"
+  - "1.1"
+  - "1.2"
```
Will this not result in a race-condition if multiple workers try to push?

DilumAluthge replied:
I tried it out. If multiple workers try to merge, the first will succeed. The subsequent workers will give an error and exit immediately. There's no timeout or anything like that.

fredrikekre replied:
Okay. I think we should run tests on multiple julia, but probably good to pick one version for the merging.

fredrikekre replied:
You can just check julia version in `.travis.yml` and exit unless it is 1.2 or something.

DilumAluthge replied:
Done!

fredrikekre commented on .test/AutoMerge/test/runtests.jl:
```diff
@@ -0,0 +1,49 @@
+                @test AutoMerge.meets_sequential_version_number(v"0.1.0", v"0.2.0")
+                @test AutoMerge.meets_sequential_version_number(v"1.0.0", v"1.0.1")
+                @test AutoMerge.meets_sequential_version_number(v"1.0.0", v"1.1.0")
+                @test AutoMerge.meets_sequential_version_number(v"1.0.0", v"2.0.0")
```
Add some tests with non-zero trailing number? E.g.
```
@test AutoMerge.meets_sequential_version_number(v"0.1.1", v"0.2.0")
@test AutoMerge.meets_sequential_version_number(v"1.0.3", v"1.1.0")
```

DilumAluthge replied:
Done!

fredrikekre commented:
This is great! 🎉  Some thoughts:
 - How about enabling it just for new versions as a start?
 - I can add a token from @zeptodoctor, or maybe we can use Kristoffer's @attobot3 . How about calling it something more specific, e.g. `GITHUB_AUTOMERGE_TOKEN`?

> Let me know if you want me to add your GitHub username to the list of usernames for which auto-merging is authorized.

Is this how it works in this PR too? e.g. that only some usernames are authorized?

DilumAluthge commented:
> How about enabling it just for new versions as a start?

Good idea, I just made that change! It is now enabled only for new versions.

> How about calling it something more specific, e.g. GITHUB_AUTOMERGE_TOKEN?

Done! The environment variable now needs to be called `GITHUB_AUTOMERGE_TOKEN`.

> Is this how it works in this PR too? e.g. that only some usernames are authorized?

Yep. Currently how I do it is I have a function `author_is_authorized`. Inside the body of the function, there is a constant `authorized_authors`, which is an array containing a list of authorized authors. As you can see [here](https://github.com/DilumAluthge/General/blob/0396b6a8886771689e83e01b90e7883210a71b7a/.test/AutoMerge/src/AutoMerge.jl#L12), currently the only authorized author is `JuliaRegistrator`, so only pull requests from @JuliaRegistrator will be considered for automatic merging. To add users to the list of authorized authors, simply modify the `authorized_authors` array.

I guess I could have put this in a config file or something, but I figured that for our use case, the only authorized user is always going to be @JuliaRegistrator.

fredrikekre commented:
>

    Is this how it works in this PR too? e.g. that only some usernames are authorized?

Yep. Currently how I do it is I have a function author_is_authorized. Inside the body of the function, there is a constant authorized_authors, which is an array containing a list of authorized authors. As you can see here, currently the only authorized author is JuliaRegistrator, so only pull requests from @JuliaRegistrator will be considered for automatic merging. To add users to the list of authorized authors, simply modify the authorized_authors array.

Ah okay, good. I misunderstood then, thought it was checking the person that invoked registrator. Can you add me as a authorized author for your test repo and I will play around a bit with it later.

DilumAluthge commented:
> Can you add me as a authorized author for your test repo and I will play around a bit with it later.

Done!

KristofferC commented:
Very nice!

DilumAluthge commented:
I've refactored the code a little bit to remove some redundant parts.

DilumAluthge commented:
@StefanKarpinski @ararslan Take a look at the code when you are able to.

@fredrikekre @KristofferC The code has changed a little bit since you last commented, so it might be helpful for you to review the code again if you have the chance.

DilumAluthge commented:
## The Big Red Button

Bot accounts on Wikipedia often have a "Big Red Button" page, which looks something like this:

![Screen Shot 2019-09-10 at 00 33 11](https://user-images.githubusercontent.com/5619885/64584474-cb258400-d362-11e9-857d-f21145d2e290.png)

If a Wikipedia administrator wants to quickly disable a bot, they click the Big Red Button.

It occurs to me that I should document the Big Red Button for the automerge functionality in this pull request.

I've added a commented-out line to the beginning of the `after_success` section of `.travis.yml`. To quickly disable all automerge functionality, uncomment the line. To re-enable the automerge functionality, re-comment the line.

This is what the `after_success` section looks like when automerging is enabled:
```yaml
after_success:
  # - exit # Administrators: uncomment this line to disable all automerging
  - if [ "$TRAVIS_JULIA_VERSION" != "1.2" ]; then exit; fi
  - julia --project=.test/AutoMerge -e 'using Pkg; Pkg.instantiate()'
  - julia --project=.test/AutoMerge -e 'using Pkg; Pkg.test("AutoMerge")' && julia --project=.test/AutoMerge -e 'using AutoMerge; automerge_travis(merge_new_packages=false, merge_new_versions=true)'
```

And this is what the `after_success` section looks like when automerging is disabled:
```yaml
after_success:
  - exit # Administrators: uncomment this line to disable all automerging
  - if [ "$TRAVIS_JULIA_VERSION" != "1.2" ]; then exit; fi
  - julia --project=.test/AutoMerge -e 'using Pkg; Pkg.instantiate()'
  - julia --project=.test/AutoMerge -e 'using Pkg; Pkg.test("AutoMerge")' && julia --project=.test/AutoMerge -e 'using AutoMerge; automerge_travis(merge_new_packages=false, merge_new_versions=true)'
```

DilumAluthge commented:
Bump. What do people think about this?

DilumAluthge commented:
Bump. Are there other thoughts on this?

fredrikekre commented:
I will try to play around with it a bit, but looks good.

One thing that we did not allow before was to drop Julia support in a patch release, is that something we want to keep enforcing?

Also, should we move this and the test-scripts into its own repository?

DilumAluthge commented:
Currently, if the automatic merging guidelines are not met, you need to look at the Travis logs to figure out why the PR wasn’t automatically merged. Should we post a comment with an explanation of the guideline that the PR didn’t meet? Or is it sufficient to have the reasons listed in the Travis logs?

tpapp commented:
If it is technically feasible, a short explanation would be nice, eg a table or list with checkboxes for all the requirements so if multiple ones fail, the user would not have to do multiple passes to figure this out.

DilumAluthge commented:
>If it is technically feasible, a short explanation would be nice, eg a table or list with checkboxes for all the requirements so if multiple ones fail, the user would not have to do multiple passes to figure this out.

@tpapp That should be feasible. I’ll work on it.

DilumAluthge commented:
@fredrikekre Do you want me to move this pull request to https://github.com/JuliaRegistries/GeneralRegistryCI.jl?

DilumAluthge commented:
I've added the "you shouldn't drop Julia support in a patch release" guideline.

I've also majorly refactored the code.

Also, I've added the "post a comment with a list of the guidelines that your PR does not meet" feature. Check out this example here of a new version PR: https://github.com/bcbi/AutoMergePrototype/pull/2

fredrikekre commented:
Great. One thing I have been thinking about is that if we do the merging from the CI bot, then PRs will basically be insta-merged. I think it would be nice to merge after ~15 minutes or so. It is quite often you realize you forgot something just after submitting the registration PR and you want to update it.

Let's discuss this more on the meeting tonight.

DilumAluthge commented:
I have updated this pull request. It now matches what we discussed in the most recent Pkg call.

heliosdrm commented:
Perhaps such a bot checking the fitness of PRs might start having the rights to approve them, and send comments telling what conditions were not met (e.g. the recurrent messages of "add [compat] to Project.toml"), rather than merging the PRs. This might already save some human effort, while mistakes of the robot would not be harmful.

DilumAluthge commented:
I'm closing this PR, and I will open an updated PR against https://github.com/JuliaRegistries/GeneralRegistryCI.jl
